cpp/src/arrow/compute/kernels/scalar_cast_string.cc

```diff
@@ -304,10 +306,32 @@ BinaryToBinaryCastExec(KernelContext* ctx, const ExecSpan& batch, ExecResult* ou
     }
   }
 
-  // Start with a zero-copy cast, but change indices to expected size
+  // Start with a zero-copy cast, but change indices to the correct size and set validity
+  // bitmap and offset if needed.
   RETURN_NOT_OK(ZeroCopyCastExec(ctx, batch, out));
-  return CastBinaryToBinaryOffsets<typename I::offset_type, typename O::offset_type>(
-      ctx, input, out->array_data().get());
+  if constexpr (sizeof(typename I::offset_type) != sizeof(typename O::offset_type)) {
+    std::shared_ptr<ArrayData> input_arr = input.ToArrayData();
+    ArrayData* output = out->array_data().get();
+
+    // Slice buffers to reduce allocation when casting the offsets buffer
```

Comment:
I think we can be more verbose about this comment - it might not be obvious for others to see how the allocation is reduced.